File: MyBakery/app/src/main/java/com/nsoft/mybakery/ProductRecyclerViewAdapter.java
package com.nsoft.mybakery;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class ProductRecyclerViewAdapter extends RecyclerView.Adapter<ProductRecyclerViewAdapter.MyViewHolder> {
    private List<Product> productList;
    private ClickListener listener;
    private LongClickListener longClickListener;

    public ProductRecyclerViewAdapter(List<Product> productList, ClickListener listener, LongClickListener longClickListener) {
        this.productList = productList;
        this.listener = listener;
        this.longClickListener = longClickListener;
    }

    public class MyViewHolder extends RecyclerView.ViewHolder {
        ImageView productImageView, productDeleteImageView;
        TextView productNameTextView, productPriceTextView;

        public MyViewHolder(@NonNull View itemView) {
            super(itemView);
            productImageView = itemView.findViewById(R.id.productImageView);
            productDeleteImageView = itemView.findViewById(R.id.productDeleteImageView);
            productNameTextView = itemView.findViewById(R.id.productNameTextView);
            productPriceTextView = itemView.findViewById(R.id.productPriceTextView);
        }
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View myView = LayoutInflater.from(parent.getContext()).inflate(R.layout.product_item_layout, parent, false);
        return new MyViewHolder(myView);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        Product product = productList.get(position);
        int id = product.getId();
        String image = product.getImage();
        String name = product.getName();
        String price = product.getPrice();

        holder.productNameTextView.setText(name);
        holder.productPriceTextView.setText("৳" + price);

        holder.productDeleteImageView.setOnClickListener(v -> {
            listener.onDeleteListener(id, position);
        });

        holder.itemView.setOnLongClickListener(v -> {
            longClickListener.onLongClickListener(id);
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return productList.size();
    }

    public interface ClickListener {
        void onDeleteListener(int id, int position);
    }

    public interface LongClickListener {
        void onLongClickListener(int position);
    }

}

File: MyBakery/app/src/main/java/com/nsoft/mybakery/adapters/ProductRecyclerViewAdapter.kt
package com.nsoft.mybakery.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.nsoft.mybakery.R
import com.nsoft.mybakery.models.Product

class ProductRecyclerViewAdapter(private val clickListener: ClickListener) :
    RecyclerView.Adapter<ProductRecyclerViewAdapter.MyViewHolder?>() {

    private var productList = ArrayList<Product>()
    fun setProductList(productList: List<Product>) {
        this.productList = productList as ArrayList<Product>
        notifyDataSetChanged()
    }

    class MyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val productImageView = itemView.findViewById<ImageView?>(R.id.productImageView)
        val productDeleteImageView = itemView.findViewById<ImageView?>(R.id.productDeleteImageView)
        val productNameTextView = itemView.findViewById<TextView?>(R.id.productNameTextView)
        val productPriceTextView = itemView.findViewById<TextView?>(R.id.productPriceTextView)

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val myView = LayoutInflater.from(parent.context)
            .inflate(R.layout.product_item_layout, parent, false)

        return ProductRecyclerViewAdapter.MyViewHolder(myView)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {

        val product = productList[position]

        holder.productNameTextView.text = product.name
        holder.productPriceTextView.text = "${product.price}"

        holder.productDeleteImageView.setOnClickListener {
            clickListener.onDeleteClickListener(product)
        }

        holder.itemView.setOnLongClickListener {
            clickListener.onItemLongClick(product)
            return@setOnLongClickListener true
        }
    }

    override fun getItemCount(): Int {
        return productList.size
    }

    interface ClickListener {
        fun onDeleteClickListener(product: Product)
        fun onItemLongClick(product: Product)
    }
}
